### ISSUE TITLE
Using OpenStreetMap as map source creates CGColorSpaceRef which are not  released IOS 6 only 

### ISSUE THREAD

**Issue by tagcat** [mapbox/DEPRECATED-mapbox-ios-sdk#194]
Using the profiler with the Mapbox example, and RMOpenStreetMapSource,  I noticed an  ever increasing number of Malloc 1.00kb object, created by colour_space_create_indexed.  Looking more closely I noted that 
when converting the 'data' to an image using  [UIImage imageWithData:tileData](in  both RMAbstractWebMapSource and RMDatabaseCache), internally a different decoding approach is used for RMMapBoxSource vrs RMOpenStreetMapSource.  With osm data, a 
CGColorSpaceRef is created which is never release!  I only see this behaviour since iOs6.  Just retested with  release xcode Version 4.6 (4H127) and iOs6.1, both simulator and Ipad4.  This is not an issue with 5.1 ! 
Can the profiler be trusted ? Can anybody else confirm that what I am seeing is real !!  If this is real, it is a serious memory leak as within a couple of minutes panning an zooming on an ipad we've lost a meg or more.

Note: I originally noticed this problem in Alpstein version two months ago (reported issue without any response)  however, I did not relealize at the time that Mapbox data was not effected.  I expect that the two are chunked differently and thus a different path taken to decode the png.

Herman

**Comment by incanus** [comment]
Assigned to [Tender discussion #759](http://support.mapbox.com/discussions/mapbox-ios-sdk/759-using-openstreetmap-as-map-source-results-in-increased-unreleased-memory-use-ios-6-only).

**Comment by tagcat** [comment]
Has anybody else tested this to confirm or refute this issue.

**Comment by incanus** [comment]
I have not tested this yet beyond minimally, where I did not see the problem that you're seeing. Unless I'm misunderstanding, this sounds like it's internal to Core Graphics and not a bug in the SDK?

**Comment by tagcat** [comment]
Hello Justin

As stated in the original bug report, this symptom does not manifest itself with Mapbox data but does show up for OSM, Yahoo and possibly other tile sources.  I belief the culprit is the PLTE chunk that exists in the OSM etc tile files.  MB tiles dont contain this chunk, uses a different decode logic and thus do not create the color_space... This only appears since ios6 not 5.  It may be a Core Graphics bug, but I find it hard to believe that no one else has seen this symptom since 6 released a half year ago.  Can find no reference at all of this problem via goggle. Maybe no one else is accessing a lot of images via the web, or the problem goes unnoticed as it doesnt show as a memory leak.  Somewhere a release is not resulting in a dealloc.  Just as a sideline, I have noticed, that releasing RMMapview does not dealloc it.  I expect that this is a side-effect of the [mapView retain] is see in RMMapTiledLayerView and RMAnnotation. Other retains may also be causing this gridlock.  
As for a solution, possibly you at Mapbox have access to apple internal bug system to determine if this problem is known to them.

Herman

**Comment by tagcat** [comment]
Hi Justin
Have spend the last couple of days trying to get a clear demonstration of the problem and try alternative contents draw method that eliminate the issue.
As stated earlier, this problem only show in tiles with indexed color table.  and not in ios5,
The order of events is as follows:  (a very abridged version)
1) Tile data read as NSData, be it web or dbcached
2) image created with ImageFromData.  Because the tile contains a palette the process uses initImagePng and CGColorSpaceCreateIndexed creates various objects including a 1k object (instruments category Malloc 1,00 KB) 
3) In drawLayer:layer:context and tileImage:drawInRect method CG takes this image and again creates a with help of ColorSpaceCreateIndexed the various objects including a 1k object as noted above.   These released almost immediatelly in drawRect process stream.
4)The ColorSpaceCreateIndexed object created in 2 never get released.

I have created a small modification to the onlineLayerViewController method in MB example that clearly demonstrates this problem.   Interestingly enough, if you DO NOT call drawInRect all objects get properly release.
I have tried various forms of converting the NSData to a CG image, including all permutations of using CG and CI images, but all of them still use ColorSpaceCreateIndexed and the problem persists.

If you like I can send the the example code, or paste it here, if you wish.
Iam sorry I keep hammering on this problem, which I now believe is a CG bug, but it is surprising how many tiles CaTileLayer uses, on a retina ipad (often 16x the tiles if at a uneven zoom level).  Within several minutes of serious panning and zooming the 1k un-released objects take more memory then actual tiles.  I also understand, being open sources, no one is responsible for a fix, I only hope a fresh look at this may result in a work around.

Herman
